core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -1293,13 +1303,50 @@ public int getRoutingNumShards() {
     /**
      * Returns the routing factor for this index. The default is <tt>1</tt>.
      *
-     * @see #getRoutingFactor(IndexMetaData, int) for details
+     * @see #getRoutingFactor(int, int) for details
      */
     public int getRoutingFactor() {
         return routingFactor;
     }
 
     /**
+     * Returns the source shard ID to split the given target shard off
+     * @param shardId the id of the target shard to split into
+     * @param sourceIndexMetadata the source index metadata
+     * @param numTargetShards the total number of shards in the target index
+     * @return a the source shard ID to split off from
+     */
+    public static ShardId selectSplitShard(int shardId, IndexMetaData sourceIndexMetadata, int numTargetShards) {
+        if (shardId >= numTargetShards) {
+            throw new IllegalArgumentException("the number of target shards (" + numTargetShards + ") must be greater than the shard id: "
+                + shardId);
+        }
+        int numSourceShards = sourceIndexMetadata.getNumberOfShards();
+        if (numSourceShards > numTargetShards) {
```

Comment:
I think we should also validate that numTargetShards is a divisor of sourceIndexMetaData.getRoutingShards() - I don't see this validation anywhere and I think this is a valid place?

Reply:
it's done in `getRoutingFactor` which is called 2 lines below

Reply:
Maybe i'm confused, but my point here is that `numTargetShards` should be a divisor of `numRoutingShards`. `getRoutingFactor` is called with `numSourceShards` & `numTargetShards` and checks that `numTargetShards` is a multiplier of `numSourceShards`. That doesn't mean it is a divisor of `numRoutingShards`? If it's not, I think we will be creating uneven shards?

Reply:
this is correct here. We do check that the number of shards is a factor of num routing shards  in may other places. I can add an assertion here but you can't even set this unless the condition holds